### PR TITLE
Federated Clients: implement MarioMetadataServiceClient

### PR DESCRIPTION
**Pull request by jonlee2**

### CODE REVIEW

File: li-apache-kafka-clients/src/main/java/com/linkedin/kafka/clients/consumer/LiKafkaFederatedConsumerImpl.java [hunk 1]
```diff
@@ -155,8 +157,12 @@ public void assign(Collection<TopicPartition> partitions) {
       return;
     }
 
-    Map<TopicPartition, ClusterDescriptor> topicPartitionToClusterMap =
-        _mdsClient.getClustersForTopicPartitions(_clientId, partitions, _mdsRequestTimeoutMs);
+    Map<TopicPartition, ClusterDescriptor> topicPartitionToClusterMap = null;
```
Comment (radai-rosenblatt) [inline li-apache-kafka-clients/src/main/java/com/linkedin/kafka/clients/consumer/LiKafkaFederatedConsumerImpl.java]: nit - =null assignment not required

File: li-apache-kafka-clients/src/main/java/com/linkedin/kafka/clients/metadataservice/MarioMetadataServiceClient.java [hunk 2]
```diff
@@ -0,0 +1,159 @@
...
+  public ClusterDescriptor getClusterForTopic(UUID clientId, String topicName, int timeoutMs)
+      throws MetadataServiceClientException {
+    Set<ClusterDescriptor> cluster = getClusterMapForTopics(clientId, new HashSet<>(Arrays.asList(topicName)),
+        timeoutMs).get(topicName);
+    if (cluster == null || cluster.isEmpty()) {
+      return null;
+    }
+    // Unless topic move is in progress, a topic must exist in only one cluster.
+    if (cluster.size() > 1) {
+      throw new IllegalStateException("topic " + topicName + " exists in more than one cluster");
```
Comment (radai-rosenblatt) [inline li-apache-kafka-clients/src/main/java/com/linkedin/kafka/clients/metadataservice/MarioMetadataServiceClient.java]: also print the clusters, and the cluster group arg used

File: li-apache-kafka-clients/src/main/java/com/linkedin/kafka/clients/metadataservice/MarioMetadataServiceClient.java [hunk 3]
```diff
@@ -0,0 +1,159 @@
...
+    // This is a temporary hack until Mario supports client registration.
+    UUID clientId = LiKafkaClientsUtils.randomUUID();
+    _clientIdToClusterGroupMap.put(clientId, clusterGroup);
+    return clientId;
+  }
+
+  @Override
+  public ClusterDescriptor getClusterForTopic(UUID clientId, String topicName, int timeoutMs)
+      throws MetadataServiceClientException {
+    Set<ClusterDescriptor> cluster = getClusterMapForTopics(clientId, new HashSet<>(Arrays.asList(topicName)),
```
[the remark this answers is not in the document]
Comment (radai-rosenblatt) [reply li-apache-kafka-clients/src/main/java/com/linkedin/kafka/clients/metadataservice/MarioMetadataServiceClient.java]: and you want to provide a cluster group name argument - same topic may exist in different groups

File: li-apache-kafka-clients/src/main/java/com/linkedin/kafka/clients/consumer/LiKafkaFederatedConsumerImpl.java (hunk 1 above)
Comment (jonlee2) [reply li-apache-kafka-clients/src/main/java/com/linkedin/kafka/clients/consumer/LiKafkaFederatedConsumerImpl.java]: done

File: li-apache-kafka-clients/src/main/java/com/linkedin/kafka/clients/metadataservice/MarioMetadataServiceClient.java [hunk 4]
```diff
@@ -0,0 +1,159 @@
...
+  }
+
+  @Override
+  public UUID registerFederatedClient(ClusterGroupDescriptor clusterGroup, Map<String, ?> configs, int timeoutMs) {
+    if (clusterGroup == null) {
+      throw new IllegalArgumentException("cluster group cannot be null");
+    }
+
+    // This is a temporary hack until Mario supports client registration.
+    UUID clientId = LiKafkaClientsUtils.randomUUID();
```
[the remark this answers is not in the document]
Comment (jonlee2) [reply li-apache-kafka-clients/src/main/java/com/linkedin/kafka/clients/metadataservice/MarioMetadataServiceClient.java]: done

File: li-apache-kafka-clients/src/main/java/com/linkedin/kafka/clients/metadataservice/MarioMetadataServiceClient.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (jonlee2) [reply li-apache-kafka-clients/src/main/java/com/linkedin/kafka/clients/metadataservice/MarioMetadataServiceClient.java]: Renamed to clusters

It takes an id of a federated client, which is for a single cluster group.

File: li-apache-kafka-clients/src/main/java/com/linkedin/kafka/clients/metadataservice/MarioMetadataServiceClient.java (hunk 2 above)
Comment (jonlee2) [reply li-apache-kafka-clients/src/main/java/com/linkedin/kafka/clients/metadataservice/MarioMetadataServiceClient.java]: done

File: li-apache-kafka-clients/src/main/java/com/linkedin/kafka/clients/metadataservice/MarioMetadataServiceClient.java [hunk 5]
```diff
@@ -0,0 +1,159 @@
...
+      }
+
+      topicPartitionToClusterMap.put(topicPartition, cluster.iterator().next());
+    }
+    return topicPartitionToClusterMap;
+  }
+
+  // For given topic names, construct a map from topic name to cluster descriptors where the topic exists. If a topic
+  // does not exist, there will be no entry for that topic in the return map.
+  private Map<String, Set<ClusterDescriptor>> getClusterMapForTopics(UUID clientId, Set<String> topicNames,
```
Comment (radai-rosenblatt) [inline li-apache-kafka-clients/src/main/java/com/linkedin/kafka/clients/metadataservice/MarioMetadataServiceClient.java]: just to clarify - this was meant to allow multiple clients (each with their own ID) to use the same MD service client?

i still think this is a bit of a violation of concerns (why should NDS client keep state per client ID? when does that state expire on client crash?)

File: li-apache-kafka-clients/src/main/java/com/linkedin/kafka/clients/metadataservice/MarioMetadataServiceClient.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (jonlee2) [reply li-apache-kafka-clients/src/main/java/com/linkedin/kafka/clients/metadataservice/MarioMetadataServiceClient.java]: Based on our discussion, added a cluster group argument.

File: li-apache-kafka-clients/src/main/java/com/linkedin/kafka/clients/metadataservice/MarioMetadataServiceClient.java (hunk 5 above)
Comment (jonlee2) [reply li-apache-kafka-clients/src/main/java/com/linkedin/kafka/clients/metadataservice/MarioMetadataServiceClient.java]: Based on our discussion, updated the code to pass cluster group to MDSClient to make it stateless.

**Pull request merged**